### ISSUE TITLE
Dynamic type not printed with  "set print object on" in .gdbinit

### ISSUE THREAD

**Issue by dannas** [microsoft/vscode-cpptools#8401]
Bug type: Debugger

**Describe the bug**
- OS and Version: Ubuntu 21.04
- VS Code Version: 1.6.2
- C/C++ Extension Version:1.7.1

The variables window shows the type as the base class when I had expected the derived class. (That's what gdb and QtCreator shows).
![image](https://user-images.githubusercontent.com/152178/141806020-0bf3897f-67d3-4ba8-95d2-87118703999e.png)

**To Reproduce**
* launch.json
```
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "g++-10 - Build and debug active file",
            "type": "cppdbg",
            "request": "launch",
            "program": "${fileDirname}/${fileBasenameNoExtension}",
            "args": [],
            "stopAtEntry": false,
            "cwd": "${fileDirname}",
            "environment": [],
            "externalConsole": false,
            "MIMode": "gdb",
            "setupCommands": [
                {
                    "description": "Enable pretty-printing for gdb",
                    "text": "-enable-pretty-printing",
                    "ignoreFailures": true
                }
            ],
            "preLaunchTask": "C/C++: g++-10 build active file",
            "miDebuggerPath": "/usr/bin/gdb"
        }
    ]
}
```
main.cpp
```
#include <iostream>

class MyBase {
    public:
        virtual int myMethod() = 0;
};

class MyDerived1 : public MyBase {
    public:
        virtual int myMethod() { return 1; }
};

class MyDerived2 : public MyBase {
    public:
        virtual int myMethod() { return 2; }
};

int main() {
    MyBase *myBase;
    MyDerived1 myDerived1;
    MyDerived2 myDerived2;
    myBase = &myDerived1;
    std::cout << myBase->myMethod() << std::endl;
    myBase = &myDerived2;
    std::cout << myBase->myMethod() << std::endl; // set breakpoint here
}
```
*Please include a code sample and `launch.json` configuration.*
0. Add "set print object on" to your .gdbinit for ability to print derived type
1. Set breakpoint on last line in main.cpp that has the text "set breakpoint here".
2. Press F5 for starting the debugger
3. Check the state of the variables window
4. Notice that myBase has type MyBase instead of the expected MyDerived2.

**Additional context**
My debug console output
```
=thread-group-added,id="i1"
GNU gdb (Ubuntu 10.1-2ubuntu2) 10.1.90.20210411-git
Copyright (C) 2021 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
=cmd-param-changed,param="auto-load safe-path",value="/"
=cmd-param-changed,param="disassembly-flavor",value="intel"
=cmd-param-changed,param="print max-symbolic-offset",value="1"
=cmd-param-changed,param="pagination",value="off"
=cmd-param-changed,param="print thread-events",value="off"
=cmd-param-changed,param="confirm",value="off"
=cmd-param-changed,param="print asm-demangle",value="on"
=cmd-param-changed,param="print object",value="on"
Warning: Debuggee TargetArchitecture not detected, assuming x86_64.
```

**Comment by Colengms** [comment]
@WardenGnaw

**Comment by WardenGnaw** [comment]
The highlighted `base` is correct, it is showing you the base type undeneath.. The actual type is visualized by the pointer value.
![image](https://user-images.githubusercontent.com/3953714/145254680-e44520f9-7afb-4296-8817-d98c97b04b96.png)

If you add member variables you can see this.
E.g.
![image](https://user-images.githubusercontent.com/3953714/145254803-08659270-770d-4aa9-932c-a6db99d5e4e0.png)
where x is from the base class and y is from the derived class.

**Comment by chrisjaquet** [comment]
Related to this, is it possible to display the name of the derived type somewhere? Something like
```
pBase: 0x7ffffffedbf0 (Derived)
  Base (base): Base
    x: 134222464
  y: 0
```
so that we can tell that `pBase` is pointing to an instance of `Derived` rather than trying to infer it from the available members?

**Comment by GekkieHenkie** [comment]
> The highlighted `base` is correct, it is showing you the base type undeneath.. The actual type is visualized by the pointer value. ![image](https://user-images.githubusercontent.com/3953714/145254680-e44520f9-7afb-4296-8817-d98c97b04b96.png)
> 
> If you add member variables you can see this. E.g. ![image](https://user-images.githubusercontent.com/3953714/145254803-08659270-770d-4aa9-932c-a6db99d5e4e0.png) where x is from the base class and y is from the derived class.

@WardenGnaw This does not help. When two different Derived types have the same member, both in name and type, you still have no idea which Concrete type is being inspected... This is not an exotic scenario. When working a with hierarchies of abstract interfaces it is very hard to keep track of which child is linked to which parent. As @chrisjaquet pointed out, showing the actual type would help tremendously.

The information is already there in GDB, as can be seen in [the post on StackOverflow by the Opening Post's author](https://stackoverflow.com/questions/69975404/vscode-c-gdb-debugger-not-showing-dynamic-type) regarding this exact issue, where gdb shows:
(gdb) p myBase
$7 = (MyBase *) 0x7fffffffdce0
# stepping
(gdb) p myBase
$8 = (MyDerived1 *) 0x7fffffffdbd0
#stepping
(gdb) p myBase
$9 = (MyDerived2 *) 0x7fffffffdbd8

In other words: the vscode plugin does not visualize debug information, which is already available in the debugger, but not yet shown on screen. Surely this cannot be 'by design'? I strongly suggest to reopen the issue and mark it as a 'bug'. Withholding useful debugging information in an IDE should never be 'by design'.

**Comment by libai6cc** [comment]
@WardenGnaw, I have the same problem and I agreed with @GekkieHenkie and @chrisjaquet, also suggest we reopen this issue.

It's very common that a project contains many dynamic class pointers. And it's really inconvenient when debugging such codes without a real type hint.

Here is a capture of QtCreator. Though `a0` is declared as `MyBase*`, it can show the real class type, which is `MyDerived*`

![2024-01-19](https://github.com/microsoft/vscode-cpptools/assets/9364643/75e46b0d-eff7-468a-a2cf-25419583c653)

Please consider our suggestion, it will be very helpful for C++ programers.
